### DISCUSSION TITLE
How to write custom code to convert parameter?

### DISCUSSION THREAD

**Discussion by DaleyKD** [ThreeMammals/Ocelot#1768] [Q&A] [answered]
## Expected Behavior / New Feature
Is it possible to execute some kind of class or code to do some logic to tweak an Upstream parameter so that it is different for a Downstream parameter?  For example, someone consuming my API may call in to the gateway with an accountID, but my downstream path may actually need to use an accountName.  The code would hook into my database service to resolve the ID to the name, and then that could be used to route for the downstream.

``` json
        {
            "DownstreamPathTemplate": "/{accountName}/api/{everything}",
            "DownstreamScheme": "http",
            "ServiceName": "my-test-service",
            "UpstreamPathTemplate": "/{accountID}/api/{everything}",
            "UpstreamHttpMethod": [],
            "AuthenticationOptions": {
                "AuthenticationProviderKey": "IdentityApiKey",
                "AllowedScopes": []
            }
        }
```

**Comment by DaleyKD** [comment]
Bump?

**Comment by ks1990cn** [comment]
Hi @DaleyKD! 
I think it will be really wrong approach to introduce database call in API Gateway to map parameters. We should optimize our gateway so we can have approx realtime communication between upstream and downstream. Introducing database call will adversely impact performance and when system increases with users it will be too slow and cost unnecessary for hitting DB everytime.

What are your thoughts on it? @DaleyKD @raman-m

**Comment by raman-m** [comment, marked as answer]
@DaleyKD  Kyle,
Are you still with Ocelot?

Ocelot doesn't transforms URL placeholder values. Path Placeholder is substituted as-is. so there is no conversions.
If you have a time to research Ocelot, use these features:
- [Delegating Handlers](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html)
- Inject `PreQueryStringBuilderMiddleware` of the [Middleware Injection](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html)

I guess, the 2nd option is the best

**Reply by DaleyKD** [reply]
Hey @raman-m ,
Thanks for the reply!
We still use Ocelot for a project at work that is currently sleeping. We've had to move over to other priorities, and hope to get back to this project next year.

We haven't touched the code in almost 2 years. It's still on .NET 5 and using Ocelot v17.0.0.

**Reply by raman-m** [reply]
Yeah, it is time to migrate to .NET 7 (and .NET 8 soon) and latest Ocelot version.

**Reply by DaleyKD** [reply]
Yup! Just not sure when we'll have time.

**Reply by RaynaldM** [reply]
migrate to 6, 7 or 8 is very simple.
don't increase your technical debt!

**Reply by raman-m** [reply]
Release: [21.0.0](../releases/tag/21.0.0)
NuGet: [21.0.0](https://www.nuget.org/packages/Ocelot#versions-body-tab)

It should be easy to migrate because Ocelot NuGet packs support multi-planforms: `net6.0` + `net7.0` + `net8.0`
